File: src/main/app/menu.ts
import { app, Menu, type BrowserWindow, type MenuItemConstructorOptions } from "electron";
import { IpcChannels } from "../../shared/ipc";

function sendCommand(window: BrowserWindow | null, command: string): void {
  window?.webContents.send(IpcChannels.appCommand, command);
}

export function createApplicationMenu(getWindow: () => BrowserWindow | null): void {
  const template: MenuItemConstructorOptions[] = [
    ...(process.platform === "darwin" ? [{ label: app.name, submenu: [{ role: "about" as const }, { type: "separator" as const }, { role: "services" as const }, { type: "separator" as const }, { role: "hide" as const }, { role: "hideOthers" as const }, { role: "unhide" as const }, { type: "separator" as const }, { role: "quit" as const }] }] : []),
    { label: "Edit", submenu: [{ role: "undo" }, { role: "redo" }, { type: "separator" }, { role: "cut" }, { role: "copy" }, { role: "paste" }, { role: "pasteAndMatchStyle" }, { role: "delete" }, { type: "separator" }, { role: "selectAll" }] },
    { label: "File", submenu: [
      { label: "Add Workspace", accelerator: "CmdOrCtrl+Shift+O", click: () => sendCommand(getWindow(), "workspace:add") },
      { label: "New Workspace View", accelerator: "CmdOrCtrl+T", click: () => sendCommand(getWindow(), "view:new") },
      { label: "Close Workspace View", accelerator: "CmdOrCtrl+W", click: () => sendCommand(getWindow(), "view:close") },
      { type: "separator" }, process.platform === "darwin" ? { role: "close" } : { role: "quit" }
    ] },
    { label: "Terminal", submenu: [
      { label: "Split Right", accelerator: "CmdOrCtrl+\\", click: () => sendCommand(getWindow(), "pane:split-right") },
      { label: "Split Down", accelerator: "CmdOrCtrl+Shift+\\", click: () => sendCommand(getWindow(), "pane:split-down") },
      { label: "Close Pane", accelerator: "CmdOrCtrl+Shift+W", click: () => sendCommand(getWindow(), "pane:close") }
    ] },
    { label: "View", submenu: [{ role: "reload" }, { role: "forceReload" }, { role: "toggleDevTools" }, { type: "separator" }, { role: "resetZoom" }, { role: "zoomIn" }, { role: "zoomOut" }, { type: "separator" }, { role: "togglefullscreen" }] }
  ];
  Menu.setApplicationMenu(Menu.buildFromTemplate(template));
}

File: src-tauri/src/menu.rs
use tauri::{
    menu::{MenuBuilder, MenuItemBuilder, PredefinedMenuItem, SubmenuBuilder},
    AppHandle, Emitter,
};

fn command_item(
    app: &AppHandle,
    command: &str,
    label: &str,
    accelerator: &str,
) -> tauri::Result<tauri::menu::MenuItem<tauri::Wry>> {
    MenuItemBuilder::with_id(command, label)
        .accelerator(accelerator)
        .build(app)
}

pub fn install_menu(app: &AppHandle) -> tauri::Result<()> {
    let add_workspace = command_item(app, "workspace:add", "Add Workspace", "CmdOrCtrl+Shift+O")?;
    let new_view = command_item(app, "view:new", "New Workspace View", "CmdOrCtrl+T")?;
    let close_view = command_item(app, "view:close", "Close Workspace View", "CmdOrCtrl+W")?;
    let split_right = command_item(app, "pane:split-right", "Split Right", "CmdOrCtrl+\\")?;
    let split_down = command_item(app, "pane:split-down", "Split Down", "CmdOrCtrl+Shift+\\")?;
    let close_pane = command_item(app, "pane:close", "Close Pane", "CmdOrCtrl+Shift+W")?;

    let edit = SubmenuBuilder::new(app, "Edit")
        .undo()
        .redo()
        .separator()
        .cut()
        .copy()
        .paste()
        .separator()
        .select_all()
        .build()?;

    let file = SubmenuBuilder::new(app, "File")
        .item(&add_workspace)
        .item(&new_view)
        .item(&close_view)
        .separator()
        .item(&PredefinedMenuItem::quit(app, None)?)
        .build()?;

    let terminal = SubmenuBuilder::new(app, "Terminal")
        .item(&split_right)
        .item(&split_down)
        .item(&close_pane)
        .build()?;

    let view = SubmenuBuilder::new(app, "View")
        .text("reload", "Reload")
        .text("force-reload", "Force Reload")
        .text("toggle-devtools", "Toggle Developer Tools")
        .separator()
        .text("reset-zoom", "Actual Size")
        .text("zoom-in", "Zoom In")
        .text("zoom-out", "Zoom Out")
        .separator()
        .text("toggle-fullscreen", "Toggle Full Screen")
        .build()?;

    let menu = MenuBuilder::new(app)
        .item(&edit)
        .item(&file)
        .item(&terminal)
        .item(&view)
        .build()?;

    app.set_menu(menu)?;
    app.on_menu_event(|app, event| {
        let command = event.id().as_ref();
        match command {
            "workspace:add" | "view:new" | "view:close" | "pane:split-right"
            | "pane:split-down" | "pane:close" => {
                let _ = app.emit("app:command", command);
            }
            _ => {}
        }
    });

    Ok(())
}
